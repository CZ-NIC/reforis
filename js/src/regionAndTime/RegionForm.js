/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

import Select from '../common/bootstrap/Select';
import {TIMEZONES, COUNTRIES} from '../utils/timezones';

RegionForm.propTypes = {
    formData: propTypes.shape({
        region: propTypes.string,
        country: propTypes.string,
        city: propTypes.string,
    }),
    setFormValue: propTypes.func.isRequired,
};


RegionForm.defaultProps = {
    formData: {},
    formErrors: {},
    setFormValue: ()=>{},
};

export default function RegionForm({formData, formErrors, setFormValue, ...props}) {
    function getRegionChoices() {
        return Object.keys(TIMEZONES)
            .sort()
            .reduce((obj, key) => {
            obj[key] = _(key);
            return obj;
        }, {})
    }

    function getCountryChoices() {
        return Object.keys(TIMEZONES[formData.region])
            .sort()
            .reduce((obj, country_code) => {
            obj[country_code] = _(COUNTRIES[country_code]);
            return obj;
        }, {})
    }

    function getCityChoices() {
        const countries = TIMEZONES[formData.region];
        const cities = countries[formData.country] || countries[Object.keys(countries)[0]];
        return Object.keys(cities)
            .sort()
            .reduce((obj, city) => {
            obj[city] = _(city);
            return obj;
        }, {})
    }

    return <>
        <h4>{_('Region settings')}</h4>
        <p>{_('Please select the timezone the router is being operated in. Correct setting is required to display ' +
            'the right time and for related functions.')}</p>
        <Select
            choices={getRegionChoices()}
            label={_('Continent or ocean')}
            value={formData.region}

            onChange={setFormValue(
                value => {
                    const country = Object.keys(TIMEZONES[value])[0];
                    const city = Object.keys(TIMEZONES[value][country])[0];
                    return {
                        region: {$set: value},
                        country: {$set: country},
                        city: {$set: city},
                    }
                }
            )}
            {...props}
        />
        <Select
            choices={getCountryChoices()}
            label={_('Country')}
            value={formData.country || ''}

            onChange={setFormValue(
                value => ({
                    country: {$set: value},
                    city: {$set: Object.keys(TIMEZONES[formData.region][value])[0]},
                })
            )}

            {...props}
        />
        <Select
            choices={getCityChoices()}
            label={_('Timezone')}
            value={formData.city || ''}

            onChange={setFormValue(
                value => ({city: {$set: value}})
            )}

            {...props}
        />
    </>
}
