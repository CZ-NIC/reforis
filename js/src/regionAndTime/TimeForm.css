#ntpServersList {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
}

#ntpServersList > p {
    margin: 0.5rem;
}
