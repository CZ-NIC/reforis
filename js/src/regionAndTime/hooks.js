/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import {useEffect, useState} from 'react';
import {useAPIPost} from '../common/APIhooks';
import API_URLs from '../common/API';

export default function useNTPDate(ws) {
    const [state, setState] = useState({
        data: null,
        id: null,
        isLoading: false
    });

    const [updateTimeState, updateTime] = useAPIPost(API_URLs.ntpUpdate);
    useEffect(() => {
        if (updateTimeState.data) {
            setState((prevState,) => ({...prevState, id: updateTimeState.data.id}))
        }
    }, [updateTimeState.data]);

    function trigger() {
        setState(state => {
            return {...state, isLoading: true}
        });
        updateTime();
    }

    useEffect(() => {
        const module = 'time';
        ws.subscribe(module)
            .bind(module, 'ntpdate_started', msg => {
                if (state.id === msg.data.id)
                    setState((prevState,) => ({...prevState, isLoading: true}));
            })
            .bind(module, 'ntpdate_finished', msg => {
                if (state.id === msg.data.id)
                    setState((prevState,) => ({
                        ...prevState,
                        isLoading: false,
                        data: {time: msg.data.time, result: msg.data.result}
                    }));
            });
    }, [state.id, ws]);

    return [state, trigger]
}
