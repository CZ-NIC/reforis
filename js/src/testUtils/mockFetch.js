/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

export default function mockFetch(data) {
    return jest.fn(() =>
        new Promise((resolve, reject) => {
            resolve({
                ok: true,
                json: () => {
                    return data ? data : {};
                },
            });
        })
    );
}
