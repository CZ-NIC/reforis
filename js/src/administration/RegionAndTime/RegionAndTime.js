/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import moment from 'moment';


import RegionForm from './RegionForm';
import TimeForm from './TimeForm';
import ForisForm from '../../formContainer/ForisForm';

import {APIEndpoints} from '../../common/API';

export default function RegionAndTime() {
    return <>
        <h3>{_('Region and time')}</h3>
        <p>{_('It is important for your device to have the correct time set. If your device\'s time is delayed, the ' +
            'procedure of SSL certificate verification might not work correctly.')}</p>
        <ForisForm
            forisConfig={{
                endpoint: APIEndpoints.regionAndTime
            }}
            prepDataToSubmit={prepDataToSubmit}
            validator={validator}
        >
            <RegionForm/>
            <TimeForm/>
        </ForisForm>
    </>
}

function validator(formData) {
    if (!moment(formData.time_settings.time).isValid())
        return {time_settings: {time: _('Time should be in YYYY-MM-DD HH:MM:SS format.')}};
    return undefined;
}

function prepDataToSubmit(formData) {
    delete formData.time_settings.ntp_servers;
    if (formData.time_settings.how_to_set_time === 'ntp')
        delete formData.time_settings.time;
    return formData
}
