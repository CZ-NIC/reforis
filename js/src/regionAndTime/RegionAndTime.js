/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import moment from "moment";

import API_URLs from "common/API";
import { ForisForm } from "foris";

import RegionForm from "./RegionForm";
import TimeForm from "./TimeForm";

RegionAndTime.propTypes = {
    ws: PropTypes.object.isRequired,
    postCallback: PropTypes.func,
};

RegionAndTime.defaultProps = {
    postCallback: () => undefined,
};

export default function RegionAndTime({ ws, postCallback }) {
    return (
        <>
            <h1>{_("Region and Time")}</h1>
            <p>
                {_(`
It is important for your device to have the correct time set. If your device's time is delayed, the
procedure of SSL certificate verification might not work correctly.
        `)}
            </p>
            <ForisForm
                forisConfig={{
                    endpoint: API_URLs.regionAndTime,
                }}
                prepDataToSubmit={prepDataToSubmit}
                postCallback={postCallback}
                validator={validator}
            >
                <RegionForm />
                <TimeForm ws={ws} />
            </ForisForm>
        </>
    );
}

function validator(formData) {
    if (!moment(formData.time_settings.time).isValid()) return { time_settings: { time: _("Time should be in YYYY-MM-DD HH:MM:SS format.") } };
    return undefined;
}

function prepDataToSubmit(formData) {
    delete formData.time_settings.ntp_servers;
    if (formData.time_settings.how_to_set_time === "ntp") delete formData.time_settings.time;
    return formData;
}
