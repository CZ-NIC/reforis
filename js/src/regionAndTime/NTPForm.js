/*
 * Copyright (C) 2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { Button, TextInput } from "foris";
import { useNTPForm } from "./hooks";

export default function NTPForm({ formData, setShown }) {
    const [formState, setFormValue, saveServer] = useNTPForm(
        formData,
        setShown
    );

    return (
        <>
            <TextInput
                label={_("NTP Server")}
                onChange={setFormValue((value) => ({
                    newServer: { $set: value },
                }))}
            />
            <Button
                onClick={saveServer}
                forisFormSize
                className="btn-primary col-lg-12 col-md-12"
            >
                {_("Save")}
            </Button>
        </>
    );
}
