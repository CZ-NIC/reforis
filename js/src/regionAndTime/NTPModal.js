/*
 * Copyright (C) 2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { Modal, ModalBody, ModalHeader } from "foris";
import NTPForm from "./NTPForm";

export default function NTPModal({
    shown,
    setShown,
    title,
    servers,
    formData,
}) {
    return (
        <Modal scrollable shown={shown} setShown={setShown}>
            <ModalHeader setShown={setShown} title={title} />
            <ModalBody>
                <NTPForm servers={servers} formData={formData} />
            </ModalBody>
        </Modal>
    );
}
