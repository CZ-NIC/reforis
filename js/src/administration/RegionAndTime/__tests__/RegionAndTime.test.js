/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {getByText, render, wait} from 'react-testing-library'

import {regionAndTime} from './__fixtures__/regionAndTime';
import mockFetch from '../../../testUtils/mockFetch';
import RegionAndTime from '../RegionAndTime';

describe('<RegionAndTime/>', () => {
    let regionAndTimeContainer;

    beforeEach(async () => {
        global.fetch = mockFetch(regionAndTime());
        const {container} = render(<RegionAndTime/>);
        await wait(() => getByText(container, 'Region and time'));
        regionAndTimeContainer = container;
    });

    it('Snapshot', async () => {
        expect(regionAndTimeContainer).toMatchSnapshot();
    });
});