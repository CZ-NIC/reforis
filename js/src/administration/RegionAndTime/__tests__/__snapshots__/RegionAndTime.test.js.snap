// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegionAndTime/> Snapshot 1`] = `
<div>
  <form>
    <h3>
      Region and time
    </h3>
    <p>
      It is important for your device to have the correct time set. If your device's time is delayed, the procedure of SSL certificate verification might not work correctly.
    </p>
    <h4>
      Region settings
    </h4>
    <p>
      Please select the timezone the router is being operated in. Correct setting is required to display the right time and for related functions.
    </p>
    <div
      class="form-group row"
    >
      <label
        class="form-control-label col-sm-3"
        for="1"
      >
        Continent or ocean
      </label>
      <div
        class="col-sm-8"
      >
        <select
          class="form-control"
          id="1"
        >
          <option
            value="Africa"
          >
            Africa
          </option>
          <option
            value="America"
          >
            America
          </option>
          <option
            value="Antarctica"
          >
            Antarctica
          </option>
          <option
            value="Arctic"
          >
            Arctic
          </option>
          <option
            value="Asia"
          >
            Asia
          </option>
          <option
            value="Atlantic"
          >
            Atlantic
          </option>
          <option
            value="Australia"
          >
            Australia
          </option>
          <option
            value="Europe"
          >
            Europe
          </option>
          <option
            value="Indian"
          >
            Indian
          </option>
          <option
            value="Pacific"
          >
            Pacific
          </option>
        </select>
        <small
          class="form-text text-muted"
        />
      </div>
    </div>
    <div
      class="form-group row"
    >
      <label
        class="form-control-label col-sm-3"
        for="2"
      >
        Country
      </label>
      <div
        class="col-sm-8"
      >
        <select
          class="form-control"
          id="2"
        >
          <option
            value="NL"
          >
            Netherlands
          </option>
          <option
            value="AD"
          >
            Andorra
          </option>
          <option
            value="GR"
          >
            Greece
          </option>
          <option
            value="RS"
          >
            Serbia
          </option>
          <option
            value="DE"
          >
            Germany
          </option>
          <option
            value="SK"
          >
            Slovakia
          </option>
          <option
            value="BE"
          >
            Belgium
          </option>
          <option
            value="RO"
          >
            Romania
          </option>
          <option
            value="HU"
          >
            Hungary
          </option>
          <option
            value="MD"
          >
            Moldova
          </option>
          <option
            value="DK"
          >
            Denmark
          </option>
          <option
            value="IE"
          >
            Ireland
          </option>
          <option
            value="GI"
          >
            Gibraltar
          </option>
          <option
            value="GG"
          >
            Guernsey
          </option>
          <option
            value="FI"
          >
            Finland
          </option>
          <option
            value="IM"
          >
            Isle of Man
          </option>
          <option
            value="TR"
          >
            Turkey
          </option>
          <option
            value="JE"
          >
            Jersey
          </option>
          <option
            value="RU"
          >
            Russia
          </option>
          <option
            value="UA"
          >
            Ukraine
          </option>
          <option
            value="PT"
          >
            Portugal
          </option>
          <option
            value="SI"
          >
            Slovenia
          </option>
          <option
            value="GB"
          >
            United Kingdom
          </option>
          <option
            value="LU"
          >
            Luxembourg
          </option>
          <option
            value="ES"
          >
            Spain
          </option>
          <option
            value="MT"
          >
            Malta
          </option>
          <option
            value="AX"
          >
            Ã…land Islands
          </option>
          <option
            value="BY"
          >
            Belarus
          </option>
          <option
            value="MC"
          >
            Monaco
          </option>
          <option
            value="NO"
          >
            Norway
          </option>
          <option
            value="FR"
          >
            France
          </option>
          <option
            value="ME"
          >
            Montenegro
          </option>
          <option
            value="CZ"
          >
            Czech Republic
          </option>
          <option
            value="LV"
          >
            Latvia
          </option>
          <option
            value="IT"
          >
            Italy
          </option>
          <option
            value="SM"
          >
            San Marino
          </option>
          <option
            value="BA"
          >
            Bosnia & Herzegovina
          </option>
          <option
            value="MK"
          >
            Macedonia
          </option>
          <option
            value="BG"
          >
            Bulgaria
          </option>
          <option
            value="SE"
          >
            Sweden
          </option>
          <option
            value="EE"
          >
            Estonia
          </option>
          <option
            value="AL"
          >
            Albania
          </option>
          <option
            value="LI"
          >
            Liechtenstein
          </option>
          <option
            value="VA"
          >
            Vatican City
          </option>
          <option
            value="AT"
          >
            Austria
          </option>
          <option
            value="LT"
          >
            Lithuania
          </option>
          <option
            value="PL"
          >
            Poland
          </option>
          <option
            value="HR"
          >
            Croatia
          </option>
          <option
            value="CH"
          >
            Switzerland
          </option>
        </select>
        <small
          class="form-text text-muted"
        />
      </div>
    </div>
    <div
      class="form-group row"
    >
      <label
        class="form-control-label col-sm-3"
        for="3"
      >
        Timezone
      </label>
      <div
        class="col-sm-8"
      >
        <select
          class="form-control"
          id="3"
        >
          <option
            value="Prague"
          >
            Prague
          </option>
        </select>
        <small
          class="form-text text-muted"
        />
      </div>
    </div>
    <h4>
      Time settings
    </h4>
    <p>
      Time should be up-to-date otherwise DNS and other services might not work properly.
    </p>
    <div
      class="form-group row"
    >
      <label
        class="form-control-label col-sm-3"
        for="4"
      >
        How to set time
      </label>
      <div
        class="col-sm-8"
      >
        <select
          class="form-control"
          id="4"
        >
          <option
            value="ntp"
          >
            Via ntp
          </option>
          <option
            value="manual"
          >
            Manually
          </option>
        </select>
        <small
          class="form-text text-muted"
        />
      </div>
    </div>
    <table
      class="table table-borderless table-hover offset-lg-3 col-lg-6 col-sm-12"
    >
      <thead>
        <tr>
          <th>
            NTP servers
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            217.31.202.100
          </td>
        </tr>
        <tr>
          <td>
            195.113.144.201
          </td>
        </tr>
        <tr>
          <td>
            195.113.144.238
          </td>
        </tr>
        <tr>
          <td>
            2001:1488:ffff::100
          </td>
        </tr>
        <tr>
          <td>
            ntp.nic.cz
          </td>
        </tr>
        <tr>
          <td>
            0.openwrt.pool.ntp.org
          </td>
        </tr>
        <tr>
          <td>
            1.openwrt.pool.ntp.org
          </td>
        </tr>
        <tr>
          <td>
            2.openwrt.pool.ntp.org
          </td>
        </tr>
        <tr>
          <td>
            3.openwrt.pool.ntp.org
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="rdt"
    >
      <div>
        <div
          class="form-group row"
        >
          <label
            class="form-control-label col-sm-3"
            for="5"
          >
            Time
          </label>
          <div
            class="col-sm-8"
          >
            <div
              class="input-group"
            >
              <input
                class="form-control "
                disabled=""
                id="5"
                type="text"
                value="2019-04-24 21:15:03"
              />
              <div
                class="input-group-append"
              >
                <a
                  class="input-group-text"
                >
                  <i
                    class="fa fa-sync-alt"
                  />
                </a>
              </div>
            </div>
            <div
              class="invalid-feedback"
            />
            <small
              class="form-text text-muted"
            />
          </div>
        </div>
      </div>
      <div
        class="rdtPicker"
      >
        <div
          class="rdtDays"
        >
          <table>
            <thead>
              <tr>
                <th
                  class="rdtPrev"
                >
                  <span>
                    â€¹
                  </span>
                </th>
                <th
                  class="rdtSwitch"
                  colspan="5"
                  data-value="3"
                >
                  April 2019
                </th>
                <th
                  class="rdtNext"
                >
                  <span>
                    â€º
                  </span>
                </th>
              </tr>
              <tr>
                <th
                  class="dow"
                >
                  Su
                </th>
                <th
                  class="dow"
                >
                  Mo
                </th>
                <th
                  class="dow"
                >
                  Tu
                </th>
                <th
                  class="dow"
                >
                  We
                </th>
                <th
                  class="dow"
                >
                  Th
                </th>
                <th
                  class="dow"
                >
                  Fr
                </th>
                <th
                  class="dow"
                >
                  Sa
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  class="rdtDay rdtOld"
                  data-value="31"
                >
                  31
                </td>
                <td
                  class="rdtDay"
                  data-value="1"
                >
                  1
                </td>
                <td
                  class="rdtDay"
                  data-value="2"
                >
                  2
                </td>
                <td
                  class="rdtDay"
                  data-value="3"
                >
                  3
                </td>
                <td
                  class="rdtDay"
                  data-value="4"
                >
                  4
                </td>
                <td
                  class="rdtDay"
                  data-value="5"
                >
                  5
                </td>
                <td
                  class="rdtDay"
                  data-value="6"
                >
                  6
                </td>
              </tr>
              <tr>
                <td
                  class="rdtDay"
                  data-value="7"
                >
                  7
                </td>
                <td
                  class="rdtDay"
                  data-value="8"
                >
                  8
                </td>
                <td
                  class="rdtDay"
                  data-value="9"
                >
                  9
                </td>
                <td
                  class="rdtDay"
                  data-value="10"
                >
                  10
                </td>
                <td
                  class="rdtDay"
                  data-value="11"
                >
                  11
                </td>
                <td
                  class="rdtDay"
                  data-value="12"
                >
                  12
                </td>
                <td
                  class="rdtDay"
                  data-value="13"
                >
                  13
                </td>
              </tr>
              <tr>
                <td
                  class="rdtDay"
                  data-value="14"
                >
                  14
                </td>
                <td
                  class="rdtDay"
                  data-value="15"
                >
                  15
                </td>
                <td
                  class="rdtDay"
                  data-value="16"
                >
                  16
                </td>
                <td
                  class="rdtDay"
                  data-value="17"
                >
                  17
                </td>
                <td
                  class="rdtDay"
                  data-value="18"
                >
                  18
                </td>
                <td
                  class="rdtDay"
                  data-value="19"
                >
                  19
                </td>
                <td
                  class="rdtDay"
                  data-value="20"
                >
                  20
                </td>
              </tr>
              <tr>
                <td
                  class="rdtDay"
                  data-value="21"
                >
                  21
                </td>
                <td
                  class="rdtDay"
                  data-value="22"
                >
                  22
                </td>
                <td
                  class="rdtDay"
                  data-value="23"
                >
                  23
                </td>
                <td
                  class="rdtDay rdtActive"
                  data-value="24"
                >
                  24
                </td>
                <td
                  class="rdtDay"
                  data-value="25"
                >
                  25
                </td>
                <td
                  class="rdtDay rdtToday"
                  data-value="26"
                >
                  26
                </td>
                <td
                  class="rdtDay"
                  data-value="27"
                >
                  27
                </td>
              </tr>
              <tr>
                <td
                  class="rdtDay"
                  data-value="28"
                >
                  28
                </td>
                <td
                  class="rdtDay"
                  data-value="29"
                >
                  29
                </td>
                <td
                  class="rdtDay"
                  data-value="30"
                >
                  30
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="1"
                >
                  1
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="2"
                >
                  2
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="3"
                >
                  3
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="4"
                >
                  4
                </td>
              </tr>
              <tr>
                <td
                  class="rdtDay rdtNew"
                  data-value="5"
                >
                  5
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="6"
                >
                  6
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="7"
                >
                  7
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="8"
                >
                  8
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="9"
                >
                  9
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="10"
                >
                  10
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="11"
                >
                  11
                </td>
              </tr>
            </tbody>
            <tfoot>
              <tr>
                <td
                  class="rdtTimeToggle"
                  colspan="7"
                >
                  21:15:03
                </td>
              </tr>
            </tfoot>
          </table>
        </div>
      </div>
    </div>
    <button
      class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
    >
       
       
      Save
    </button>
  </form>
</div>
`;
